File: leetcode/07-Dynamic-Programming/Grid DP/2684. Maximum Number of Moves in a Grid.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxMoves(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n));
        for (int i = 0; i < m; ++i)
        {
            dp[i][0] = 1;
        }
        for (int j = 0; j < n; ++j)
        {
            for (int i = 0; i < m; ++i)
            {
                if (j + 1 < n)
                {
                    if (i - 1 >= 0 && grid[i - 1][j + 1] > grid[i][j])
                    {
                        dp[i - 1][j + 1] |= dp[i][j];
                    }
                    if (grid[i][j + 1] > grid[i][j])
                    {
                        dp[i][j + 1] |= dp[i][j];
                    }
                    if (i + 1 < m && grid[i + 1][j + 1] > grid[i][j])
                    {
                        dp[i + 1][j + 1] |= dp[i][j];
                    }
                }
                
            }
        }
        int ans = 0;
        for (int j = n - 1; j >= 0; --j)
        {
            for (int i = 0; i < m; ++i)
            {
                if (dp[i][j] > 0)
                {
                    ans = j;
                    return ans;
                }
            }
        }
        return ans;
    }
};

/*
    还好
    注意是列历遍，不要行历遍
*/

File: leetcode/05-Grid-DFS/Grid DFS/go/2684. Maximum Number of Moves in a Grid.go
package GridDFS

func maxMoves(grid [][]int) int {
	dirs := [][]int{{0, 1}, {-1, 1}, {1, 1}}
	m, n := len(grid), len(grid[0])
	mem := make([][]int, m)
	for i := range mem {
		mem[i] = make([]int, n)
		for j := range mem[i] {
			mem[i][j] = -1
		}
	}
	var dfs func(i, j int) int
	dfs = func(i, j int) int {
		if mem[i][j] != -1 {
			return mem[i][j]
		}
		res := 0
		for _, dir := range dirs {
			x := i + dir[0]
			y := j + dir[1]
			if x < 0 || x >= m || y < 0 || y >= n || grid[x][y] <= grid[i][j] {
				continue
			}
			res = max(res, 1+dfs(x, y))
		}
		mem[i][j] = res
		return res
	}
	ans := 0
	for i := 0; i < m; i++ {
		ans = max(ans, dfs(i, 0))
	}
	return ans
}
